Dockerfile

```diff
@@ -1,17 +1,17 @@
 # set base image (host OS)
-FROM python:3.8
+FROM jfloff/alpine-python:3.8-slim
 
 # set the working directory in the container
 WORKDIR /code
 
 # copy the dependencies file to the working directory
 COPY requirements.txt .
 
-# install dependencies
-RUN pip install -r requirements.txt
-
 # copy the content of the local src directory to the working directory
 COPY main.py .
 
-# command to run on container start
-CMD [ "python", "./main.py" ]
+# -P is part of the base image
+# https://github.com/jfloff/alpine-python/blob/7ae8be3ff086e91051a478ae5fa9acb9f4ec6967/3.8-slim/entrypoint.sh#L23
+# -P: requirements.txt file location (in the container), default is /requirements.txt
+# The base image attempts to pip install dependencies when the container starts using requirements.txt
+CMD [ "-P", "/code/requirements.txt", "python", "./main.py" ]
```

Comment:
If a dependency changes upstream you containers will not be reproducable. You should look into something like https://pypi.org/project/xar/ and build self contained executables.

Reply:
`xar` sounds interesting. In general It would be better to build the app and its deps into a container layer. I tried to do so but hit some pip error likely due to alpine. Just didn't feel like troubleshooting that and this was enough of a win to merge

Reply:
I 100% disagee its better to to build or install the deps at container time. It costs time at image build which will reduce the ephermality of the image (i.e rolling containers takes more time so you do it less). Also it breaks reproducability an image built yesterday will not be the same as today if somthing upstream gets comprimised or broken. Third by bundling together with like a xar you can gaurentee the deps you need dont change between test and prod.

Reply:
Also you can update the environment without updating the packages, what happens when default python moves to 3.9 but this has some hard dep on 3.8 lib